### PR TITLE
feat(npm): add hook for getting the authentication header

### LINKED ISSUE #2067: [Feature] Add the ability to build authentication plugins
- [x] I'd be willing to implement this feature
- [ ] This feature can already be implemented through a plugin

**Describe the user story**

When dealing with 3rd party private repos it would be nice to have the ability to make an authentication plugin to avoid the need for hard coded credentials

**Describe the solution you'd like**

At the most basic level you would just need to add a hook to npmHttpUtils.getAuthenticationHeader which would allow you to change the authentication headed. The plugin would likely only need the same arguments which are passed to getAuthConfiguration. It would then just return the authentication header string.

**Describe the drawbacks of your solution**

There is a fair amount of potential for a malicious extension to break things so the error messages for failures here will likely need to be pretty robust.

**Describe alternatives you've considered**

You could also provide a hook instead in getAuthConfiguration to give the opportunity to override just the auth config. Allowing replacement of the whole string passed via the header is more flexible though since not all authentication systems use Bearer or Basic auth

**Additional context**

If this were implemented I would likely build an Azure DevOps auth plugin

### PR DESCRIPTION
**Pull request by waterfoul**
**What's the problem this PR addresses?**

Adds a hook for adding authentication headers, intended to be used for OAuth style authentication flows.
Closes #2067

...

**How did you fix it?**

I used reduceHook inside getAuthenticationHeader to collect the potential values from the hook, then return the provided value if there is one instead of checking npmAuthToken or npmAuthIdent

...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by waterfoul** [comment]
@merceyz / @arcanis any chance one of you could take a look at this?

**Comment by waterfoul** [comment]
@merceyz or @arcanis is there anything I should be doing differently to get this reviewed?

**Comment by waterfoul** [comment]
@arcanis Is there something wrong with this PR? If there is I'd like to fix it

File: .yarn/versions/c743d681.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+releases:
+  "@yarnpkg/plugin-npm": minor
+
+declined:
+  - "@yarnpkg/plugin-compat"
+  - "@yarnpkg/plugin-npm-cli"
+  - "@yarnpkg/cli"
```
Comment (arcanis) [inline .yarn/versions/c743d681.yml:7]: You need to mark `@yarnpkg/cli` as minor as well.

File: packages/plugin-npm/sources/npmHttpUtils.ts [hunk 2]
```diff
@@ -178,13 +179,21 @@ export async function del(path: string, {attemptedAs, configuration, headers, id
...
     return null;
 
+  const header = await configuration.reduceHook((hooks: Hooks) => {
+    return hooks.getNpmAuthenticationHeader;
+  }, undefined, registry, {configuration, ident});
+
+  if (header)
+    return header;
+
+
```
Comment (arcanis) [inline packages/plugin-npm/sources/npmHttpUtils.ts:196]:
```suggestion

```

**Review by arcanis — COMMENTED** [review]
Hey @waterfoul! Sorry for the delay, it fell off my radar 🙁

Just two nits (and a rebase); I'd have done it myself but your PR doesn't seem open to maintainers edit (probably because of your org settings).

File: packages/plugin-npm/sources/npmHttpUtils.ts [hunk 3]
```diff
@@ -178,13 +179,21 @@ export async function del(path: string, {attemptedAs, configuration, headers, id
...
+async function getAuthenticationHeader(registry: string, {authType = AuthType.CONFIGURATION, configuration, ident}: {authType?: AuthType, configuration: Configuration, ident: RegistryOptions['ident']}) {
   const effectiveConfiguration = npmConfigUtils.getAuthConfiguration(registry, {configuration, ident});
   const mustAuthenticate = shouldAuthenticate(effectiveConfiguration, authType);
 
   if (!mustAuthenticate)
     return null;
 
+  const header = await configuration.reduceHook((hooks: Hooks) => {
+    return hooks.getNpmAuthenticationHeader;
+  }, undefined, registry, {configuration, ident});
```
Comment (merceyz) [inline packages/plugin-npm/sources/npmHttpUtils.ts:191]: Would this be a good time to use an object instead of multiple arguments?

**Comment by arcanis** [comment]
I forgot to wait for your fixes 🤦‍♀️ Will fix it on master, nevermind.

**Pull request merged**